src/worker/triggers/keyword.ts

```diff
@@ -71,40 +74,45 @@ const Defaults = {
  *     in the range [0, 1]
  */
 export default class KeywordRecognizer implements SpeechProcessor {
-  private models?: CommandModels
+  private config: Required<KeywordTriggerConfig>
+  private models: CommandModels
   private hopSamples: number
-  private melWidth: number
-  private sampleWindow: RingBuffer<number> = new RingBuffer<number>(0)
-  private encodeWindow: RingBuffer<Tensor> = new RingBuffer<Tensor>(0)
-  private encodeState: Tensor = tf.zeros([1])
-  private frameWindow: RingBuffer<Tensor> = new RingBuffer<Tensor>(0)
-  private keywordClasses: string[]
-  private threshold: number
+  private sampleWindow = new RingBuffer<number>(0)
+  private encodeWindow = new RingBuffer<Tensor>(0)
+  private encodeState = self.tf.zeros([1])
+  private frameWindow = new RingBuffer<Tensor>(0)
   private vadActive = false
 
-  constructor(config: SpeechConfig) {
-    // number of samples to slide the window for each filter run
-    this.hopSamples = (config.hopLength || Defaults.hopLength) * (config.sampleRate / 1000)
-    this.melWidth = config.melWidth || Defaults.melWidth
-    // if (classes.length != expected) {
+  static async create(config: SpeechConfig) {
+    if (!config.baseKeywordUrl) {
+      throw new Error('keyword URL required')
+    }
     if (!config.keywordClasses || config.keywordClasses.length == 0) {
       throw new Error('keyword classes required')
-      // return new Error(`invalid number of keyword classes: ${classes.length} != ${expected}`)
     }
-    this.keywordClasses = config.keywordClasses
-    this.threshold = config.keywordThreshold || Defaults.threshold
+    config.sampleRate = Number(config.sampleRate)
+    if (isNaN(config.sampleRate)) {
+      throw new Error('sampleRate is required')
+    }
+    const models = await KeywordRecognizer.loadModels(config.baseKeywordUrl, config.sampleRate)
```

Comment:
The main thing I was trying to avoid with the separation of constructor and async `init` for the individual stages was blocking on this model load during construction of the pipeline. I'd rather defer that to `start` like the other platforms do, but maybe it's not that big of a deal. I guess this allows the demo to start quicker.

Reply:
I just thought it was an artifact leftover from OO lands. I'm not a fan of passing around constructors in JS. I do see the merit in separating pipeline construction from pipeline initialization. However, what I've done here is require the models to  be loaded before constructing the pipeline, (`loadModels` does not need a pipeline to load models). The `create` functions are just convenience methods that combine loading the models and constructing the processor. Whether the pipeline actually starts can and should still be separate, and I may not have done that part right.